<?xml version="1.0" encoding="UTF-8"?>
<root BTCPP_format="4"
      main_tree_to_execute="MainTree">
  <BehaviorTree ID="MainTree">
    <Sequence name="GetStatusThenPrioritize">
      <SubRobotStatus topic_name="robot_status"
                      robot_status="{robot_status}"/>
      <Fallback name="PrioritizeAttackOrNormalLogic">
        <Sequence name="ExecuteAttackLogic">
          <IsAttaked message="{robot_status}"/>
          <Fallback name="AttackedHPBasedAction">
            <Sequence name="AttackedGoHomeLowHP">
              <Inverter>
                <IsStatusOK message="{robot_status}"
                            hp_threshold="101"
                            blue_outpost_hp_threshold="0"
                            red_outpost_hp_threshold="0"
                            heat_threshold="9999"/>
              </Inverter>
              <SendGoal goal_pose="-0.35;-7.77;0; 0;0;0;1"/>
            </Sequence>
            <SendGoal name="AttackedGoToAggressivePos"
                      goal_pose="-2.94;3.43;0; 0;0;0;1"/>
          </Fallback>
        </Sequence>
        <Sequence name="OriginalComplexLogic">
          <SubGameStatus topic_name="game_status"
                         game_status="{game_status}"/>
          <Fallback>
            <Sequence name="HandleGameNotStarted">
              <Inverter>
                <IsGameTime message="{game_status}"
                            game_progress="4"
                            lower_remain_time="0"
                            higher_remain_time="420"/>
              </Inverter>
              <RobotControl stop_gimbal_scan="true"
                            chassis_spin_vel="0.0"/>
            </Sequence>
            <Sequence name="HandleGameInProgress">
              <IsGameTime message="{game_status}"
                          game_progress="4"
                          lower_remain_time="0"
                          higher_remain_time="420"/>
              <RateController hz="10">
                <Fallback name="StatusBasedMovement">
                  <Sequence name="LowHPGoSafePoint">
                    <Inverter>
                      <IsStatusOK message="{robot_status}"
                                  hp_threshold="161"
                                  blue_outpost_hp_threshold="0"
                                  red_outpost_hp_threshold="0"
                                  heat_threshold="9999"/>
                    </Inverter>
                    <SendGoal goal_pose="3.64;1.36;0; 0;0;0;1"/>
                  </Sequence>
                  <Sequence name="HighHPGoAggressivePoint">
                    <IsStatusOK message="{robot_status}"
                                hp_threshold="301"
                                blue_outpost_hp_threshold="0"
                                red_outpost_hp_threshold="0"
                                heat_threshold="9999"/>
                    <SendGoal goal_pose="-7.63;5.60;1; 0;0;0;1"/>
                  </Sequence>
                  <SendGoal name="DefaultOrMediumHPGoInitialPoint"
                            goal_pose="-7.63;5.60;1; 0;0;0;1"/>
                </Fallback>
              </RateController>
            </Sequence>
          </Fallback>
        </Sequence>
      </Fallback>
    </Sequence>
  </BehaviorTree>

  <!-- Description of Node Models (used by Groot) -->
  <TreeNodesModel>
    <Action ID="IsAttaked"
            editable="true">
      <input_port name="message"
                  default="{robot_status}"/>
    </Action>
    <Action ID="IsGameTime"
            editable="true">
      <input_port name="message"
                  default="{game_status}"/>
      <input_port name="game_progress"
                  default="4"/>
      <input_port name="lower_remain_time"
                  default="0"/>
      <input_port name="higher_remain_time"
                  default="420"/>
    </Action>
    <Action ID="IsStatusOK"
            editable="true">
      <input_port name="message"
                  default="{robot_status}"/>
      <input_port name="hp_threshold"
                  default="0"/>
      <input_port name="blue_outpost_hp_threshold"
                  default="0"/>
      <input_port name="red_outpost_hp_threshold"
                  default="0"/>
      <input_port name="heat_threshold"
                  default="9999"/>
    </Action>
    <Decorator ID="RateController">
      <input_port name="hz">Rate</input_port>
    </Decorator>
    <Action ID="RobotControl"
            editable="true">
      <input_port name="stop_gimbal_scan"
                  default="True"/>
      <input_port name="chassis_spin_vel"
                  default="0.0"/>
    </Action>
    <Action ID="SendGoal"
            editable="true">
      <input_port name="goal_pose"
                  default="0;0;0; 0;0;0;1"/>
    </Action>
    <Action ID="SubGameStatus"
            editable="true">
      <input_port name="topic_name"
                  default="game_status"/>
      <output_port name="game_status"
                   default="{game_status}"/>
    </Action>
    <Action ID="SubRobotStatus"
            editable="true">
      <input_port name="topic_name"
                  default="robot_status"/>
      <output_port name="robot_status"
                   default="{robot_status}"/>
    </Action>
  </TreeNodesModel>

</root>
